### PR TITLE
nixos-rebuild-ng: implement `--target-host`

### PR DESCRIPTION
**Pull request by thiagokokada**
This PR ended up bigger than I expected because it needed lots of changes to support remote builders. Why this took so much code? Well:
- Need to differentiate between the multiple possible Nix flags, e.g.: `--use-substitutes` is only for `nix-copy-closure` while `--include` is only for `nix-build`
- Need an abstraction to run command remotely, so now we have `process.run_wrapper`
- We need some things done completely different remotely, e.g.: get hostname

However since everything is here now, it should be easy to add `--build-host` next. But `--build-host` is sufficiently different that I want to do this in a separate PR, e.g.: we can't use `nix-build` and need to use `nix-copy-closure` and `nix-store -r` to  realise the store path.

~~This is basically a bug to bug implementation of the original. There is an issue with `--use-remote-sudo` asking for password twice that is also reproducible in the original. It can be fixed, however I don't think it is worth it considering that we eventually we will have `activate` script once it gets back into the tree, and once we have `activate` this issue goes away.~~

I changed the implementation of remote to not depend in `ssh -t` (pseudo TTY) anymore. This is a headache since it changes lots of behaviors inside SSH (for example, it redirects all the output to stdout), and it is also unreliable.

However, `sudo` depends in the TTY to ask for password, so now we have a flag `--ask-sudo-password` that will ask the password before running the tasks remotely. For local this doesn't change anything.

I am not sure if there is some concerns about security here: we are storing the password in memory and there is no good way in Python to remove it (like `memset` that I assume `sudo` uses here), so technically it will stay there until some program overwrites it. I don't think it is a huge concern though, people store secrets in environment variables all the time and this is probably easier to dump than memory. Also, if someone wants to grab your sudo password and have access to your computer to do so, they could run a keylogger and it would be much easier than dumping your memory. But if you are a reviewer and have some concerns here, please say so. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by thiagokokada** [comment]
CC @Scrumplex for review.

**Comment by thiagokokada** [comment]
@ofborg build nixos-rebuild-ng.tests

File: pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/utils.py [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
 
-def dict_to_flags(d: dict[str, Any]) -> list[str]:
+def dict_to_flags(d: dict[str, Args]) -> list[str]:
     flags = []
     for key, value in d.items():
         flag = f"--{'-'.join(key.split('_'))}"
         match value:
-            case None | False:
+            case None | False | 0 | []:
                 pass
```
Comment (lucasew) [inline pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/utils.py:16]: continue

File: pkgs/by-name/ni/nixos-rebuild-ng/src/tests/test_models.py [hunk 2]
```diff
@@ -98,3 +111,37 @@ def test_profile_from_name(mock_mkdir: Any) -> None:
         Path("/nix/var/nix/profiles/system-profiles/something"),
     )
     mock_mkdir.assert_called_once()
+
+
+def test_remote_from_name(monkeypatch: Any, tmpdir: Path) -> None:
+    monkeypatch.setenv("NIX_SSHOPTS", "")
+    assert m.Remote.from_arg("user@localhost", None, tmpdir) == m.Remote(
+        "user@localhost",
```
Comment (lucasew) [inline pkgs/by-name/ni/nixos-rebuild-ng/src/tests/test_models.py:119]: use keyword arguments to make it explicit

File: pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/utils.py (hunk 1 above)
Comment (thiagokokada) [reply pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/utils.py:16]: Done.

**Comment by thiagokokada** [comment]
@ofborg build nixos-rebuild-ng.tests

**Comment by thiagokokada** [comment]
@ofborg build nixos-rebuild-ng nixos-rebuild-ng.tests

File: pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/__init__.py [hunk 3]
```diff
@@ -112,7 +112,7 @@ def parser_warn(msg):
 
     # TODO: use deprecated=True in Python >=3.13
     if args.no_ssh_tty:
-        parser_warn("--no-ssh-tty deprecated, SSH's pseudo-TTY is never used anymore")
+        parser_warn("--no-ssh-tty deprecated, SSH's TTY is never used anymore")
```
Comment (Scrumplex) [inline pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/__init__.py:115]: Isn't it technically still used for the sudo prompt? Maybe reword this:

`--no-ssh-tty deprecated, a TTY is only allocated if it's explicitly needed`

Comment (thiagokokada) [reply pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/__init__.py:115]: No, SSH's TTY (the `-t` flag) is never used anymore, this is why we have `--ask-sudo-password` (that asks the password using Python's `getpass` instead and pass the password via stdin using `--stdin` flag in `sudo`).

The reason for it is because SSH's TTY mess up with lots of the expectations you have with SSH, e.g.: it combines stdout and stderr in one stream (stdout), so if you expect a command to print something to stderr for example it will be wrong.

Of course this can be workaround, but it is better to simple not do it. The bonus is that using something like `--ask-sudo-password` means we avoid the previous issue where the `sudo` password were asked twice because each SSH connection created a new TTY and by default the SSH cache is associated to each TTY (it can be changed, but it is rarely done so). Also this is the way other tools do it, for example Ansible has a `--ask-become-pass` that is similar behavior of `--ask-sudo-password`.

Comment (Scrumplex) [reply pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/__init__.py:115]: Oh, I see. I misread the code earlier.

**Comment by thiagokokada** [comment]
Pushed a commit to simplify how the Nix arguments are declared. Instead of needing to group them twice, now we are using the Argparse groups to group the arguments. This avoid declaration of the groups twice.

**Comment by thiagokokada** [comment]
Will stop forcing push for now, I am doing that bad habit of just rewriting code since refactoring is fun 😅.

CC @Scrumplex for review.

**Comment by lucasew** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359097`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>nixos-rebuild-ng</li>
    <li>nixos-rebuild-ng.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359097`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>nixos-rebuild-ng</li>
    <li>nixos-rebuild-ng.dist</li>
  </ul>
</details>

**Review by Scrumplex — COMMENTED** [review]
Changes LGTM. I will test this and report back in a moment

**Review by Scrumplex — APPROVED** [review]
